controller/utils.go

```diff
@@ -36,19 +36,87 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-func replacePodsAndWait(
+// SegmentAndReplacePods acts when a scheduler rolling update is needed.
+// It segment the list of current pods in chunks of size maxSurge and replace them with new ones
+func SegmentAndReplacePods(
+	ctx context.Context,
 	logger logrus.FieldLogger,
 	roomManager models.RoomManager,
 	mr *models.MixedMetricsReporter,
 	clientset kubernetes.Interface,
 	db pginterfaces.DB,
 	redisClient redisinterfaces.RedisClient,
 	willTimeoutAt time.Time,
+	configYAML *models.ConfigYAML,
+	pods []v1.Pod,
+	scheduler *models.Scheduler,
+	operationManager *models.OperationManager,
+	maxSurge int,
 	clock clockinterfaces.Clock,
+) (timeoutErr, cancelErr, err error) {
+	schedulerName := scheduler.Name
+	l := logger.WithFields(logrus.Fields{
+		"source":    "SegmentAndReplacePods",
+		"scheduler": schedulerName,
+	})
+
+	// segment pods in chunks
+	podChunks := segmentPods(pods, maxSurge)
+
+	for i, chunk := range podChunks {
+		l.Debugf("updating chunk %d: %v", i, names(chunk))
+
+		// replace chunk
+		timedout, canceled, errored := replacePodsAndWait(
+			l,
+			roomManager,
+			mr,
+			clientset,
+			db,
+			redisClient,
+			willTimeoutAt,
+			configYAML,
+			chunk,
+			scheduler,
+			operationManager,
+			clock,
+		)
+
+		if timedout {
+			timeoutErr = errors.New("timedout waiting rooms to be replaced, rolled back")
+			l.WithError(err).Error(timeoutErr.Error())
+			break
+		}
+
+		if canceled {
+			cancelErr = errors.New("operation was canceled, rolled back")
+			l.WithError(err).Error(cancelErr.Error())
+			break
+		}
+
+		if errored != nil {
+			err = errored
+			l.WithError(err).Error(err.Error())
```

Comment:
Here you are duplicating the message information. Prefer `l.WithError(err).Error("some tag string")` to be easier to find in the project.